lib/fugue.rego

```diff
@@ -37,39 +37,52 @@ resources(rt) = ret {
 }
 
 allow_resource(resource) = ret {
-  ret = {
+  ret := allow({"resource": resource})
+}
+
+allow(params) = ret {
+  ret := {
     "valid": true,
-    "id": resource.id,
-    "message": "",
-    "type": resource._type,
-    "provider": resource._provider,
+    "id": params.resource.id,
+    "type": params.resource._type,
+    "message": object.get(params, "message", ""),
+    "provider": params.resource._provider,
   }
 }
 
 deny_resource(resource) = ret {
-  ret = deny_resource_with_message(resource, "")
+  ret := deny({"resource": resource})
 }
 
 deny_resource_with_message(resource, message) = ret {
-  ret = {
+  ret := deny({"resource": resource, "message": message})
+}
+
+deny(params) = ret {
+  ret := {
     "valid": false,
-    "id": resource.id,
-    "message": message,
-    "type": resource._type,
-    "provider": resource._provider,
+    "id": params.resource.id,
+    "type": params.resource._type,
+    "message": object.get(params, "message", ""),
+    "attribute": object.get(params, "attribute", null),
+    "provider": params.resource._provider,
   }
 }
 
 missing_resource(resource_type) = ret {
-  ret = missing_resource_with_message(resource_type, "")
+  ret := missing({"resource_type": resource_type})
 }
 
 missing_resource_with_message(resource_type, message) = ret {
-  ret = {
+  ret := missing({"resource_type": resource_type, "message": message})
+}
+
+missing(params) = ret {
+  ret := {
     "valid": false,
     "id": "",
-    "message": message,
-    "type": resource_type,
+    "type": params.resource_type,
+    "message": object.get(params, "message", "invalid"),
```

Comment:
It won't let me comment on the line since it wasn't changed in this PR, but would it make sense to get rid of the `TODO` below and change line 89?:
```
"provider": object.get(params, "provider", ""),
```